Fixes for arm64 cross compile

glevand opened the pull request:
This series gets gimme's arm64 cross compile support working correctly.  It also has a few other general enhancements.

I tested this series using CoreOS Ignition, which needs cgo support: https://travis-ci.org/glevand/coreos--ignition/builds.

glevand commented:
Fixed OSX build errors, added arm64 travis tests.

meatballhat commented on gimme:
```diff
@@ -164,38 +169,53 @@ _extract() {
+		export CC_FOR_TARGET="${GIMME_CC_FOR_TARGET}"
+
+		if [ "${GIMME_DEBUG}" = "2" ]; then
+			./make.bash 1>&2 || return 1
```
Can this `tee $make_log` instead of only going to stdout?  That way the log is present for any value of `GIMME_DEBUG`.

meatballhat commented on gimme:
```diff
@@ -360,6 +384,39 @@ _assert_version_given() {
 : ${GIMME_TYPE:=auto} # 'auto', 'binary', 'source', or 'git'
 : ${GIMME_BINARY_OSX:=osx10.8}
 
+while [[ $# -gt 0 ]]; do
```
👍

meatballhat commented:
:+1: :heart: lovely!  Just one question 😸

glevand committed:
travis: Spilt into matrix

Split the processing into a  matrix based on the build target.

To avoid errors like these:

  The job exceeded the maxmimum time limit for jobs, and has been terminated.
  Can not write: No space left on device

Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Add GIMME_HOSTOS

For cross build support.

Signed-off-by: Geoff Levand <geoff@infradead.org>

lupan2005 committed:
gimme: Add version, arch params to _try_binary

To support installing a binary bootstrap compiler to be used for
building a cross compiler add "arch" and "version" parameters
to the _binary and _try_binary functions and replace any use of
${GIMME_ARCH} and ${GIMME_GO_VERSION} in those routines with the
passed parameters.

Also, to enhance readability, convert _binary and _try_binary to use
local variables for their passed parameters.

Signed-off-by: lupan2005 <lupan2005@gmail.com>
[Cleaned up commit message.]
Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Verbose go make

To aid in debugging go build problems, redirect the go make output
to stderr when GIMME_DEBUG=2 is set.

Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Move code to _setup_bootstrap

To make way for expanding the bootstrap processing pull that
code out into its own function _setup_bootstrap.  Just moves
the code, no functional change.

Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Add arm64 version warning

arm64 support was added in go1.5.  Print an error messsage and fail
if an older version is requested.

Signed-off-by: lupan2005 <lupan2005@gmail.com>
[Split off from larger patch and cleaned up.]
Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Fix bootstrap not found message

Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Process all flags

Add a loop around the flag processing so that multiple flags will not
result in running with erroneous parameters.

Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Try binary bootstrap

If an existing bootstrap compler is not found see if a binary
version is avalable.  The exising code would build from git if
no existing was found.

Signed-off-by: lupan2005 <lupan2005@gmail.com>
[Split off from larger patch and cleaned up.]
Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Add GIMME_CC_FOR_TARGET

To allow for control of cgo support. Also, add a default value for
arm64 linux builds.

Signed-off-by: lupan2005 <lupan2005@gmail.com>
[Split off from larger patch]
Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Move flag processing up

To allow for other cleanups.  Adds a local variable 'force' which is set
when the -f flag is passed.

Signed-off-by: lupan2005 <lupan2005@gmail.com>
[Split off from larger patch]
Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
gimme: Add GIMME_CGO_ENABLED

To allow for control of cgo support.

Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand committed:
travis: Add arm64 CI tests

Signed-off-by: Geoff Levand <geoff@infradead.org>

glevand commented:
Rebased to latest, added tee in go verbose output.

meatballhat commented on gimme:
```diff
@@ -169,38 +174,53 @@ _extract() {
 		local make_log="${1}/make.${GOOS}.${GOARCH}.log"
-		./make.bash &> $make_log || return 1
+		if [ "${GIMME_DEBUG}" > "1" ]; then
+			./make.bash 2>&1 | tee $make_log 1>&2 || return 1
```
:clap: 🎺

meatballhat commented:
@glevand Much thanks! 😻 🎷 🐻
